tests/importer/importer.cpp

```diff
@@ -0,0 +1,761 @@
+    rightSourceModel->addComponent(libcellml::Component::create("c"));
+    rightSourceModel->addComponent(libcellml::Component::create("d"));
+
+    // Add a model manually to the library, including the URL that it will replace in future imports.
+    EXPECT_TRUE(importer->addModel(wrongSourceModel, "i_dont_exist.cellml"));
+    // Can't add to the same URL key more than once.
+    EXPECT_FALSE(importer->addModel(rightSourceModel, "i_dont_exist.cellml"));
+
+    // Use replaceModel instead.
+    EXPECT_FALSE(importer->replaceModel(rightSourceModel, "not_in_library"));
+    EXPECT_TRUE(importer->replaceModel(rightSourceModel, "i_dont_exist.cellml"));
+
+    importer->resolveImports(model, "");
+
+    EXPECT_EQ(size_t(0), importer->issueCount());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, getNonexistentModel)
+{
+    auto importer = libcellml::Importer::create();
+    auto model = libcellml::Model::create("model");
+    importer->addModel(model, "howdy");
+    EXPECT_EQ(nullptr, importer->library("bonjour"));
+}
+
+TEST(Importer, getListOfDependencies)
+{
+    // This test shows how a model can be interrogated to return a list of its dependencies/URLs.
+    // These then become the URL "keys" that must be supplied to the importer, if their actual location
+    // should not be used or is inaccessible.
+
+    // Note that in order to resolve the dependencies further than the first generation, the models will
+    // need to be instantiated and saved anyway.  The external dependencies are those which, during the
+    // resolution for this model, were sourced from locations outside the library.  They will have since
+    // been saved to the importer library, and subsequent calls to resolveImports for those same imported
+    // files will not change the dependencies list.
+
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("importer/diamond.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    EXPECT_EQ(size_t(3), importer->externalDependencyCount());
+    for (size_t e = 0; e < importer->externalDependencyCount(); ++e) {
+        auto info = importer->externalDependency(e);
+        EXPECT_NE(nullptr, importer->library(info.first));
+    }
+    std::pair<std::string, std::string> empty;
+    EXPECT_EQ(empty, importer->externalDependency(999));
+}
+
+TEST(Importer, library)
+{
+    auto importer = libcellml::Importer::create();
+    auto model1 = libcellml::Model::create("model1");
+    auto model2 = libcellml::Model::create("model2");
+    auto model3 = libcellml::Model::create("model3");
+    auto model4 = libcellml::Model::create("model4");
+
+    EXPECT_TRUE(importer->addModel(model1, "a_model_the_first.cellml"));
+    EXPECT_TRUE(importer->addModel(model2, "b_tweede_model.cellml"));
+    EXPECT_TRUE(importer->addModel(model3, "c_troisieme_model.cellml"));
```

Comment:
Typos: it should be `c_troisième_modèle.cellml`. :)

Reply:
Haha! I didn't translate the "model" bit ... otherwise the Dutch would be "monster" (I think ...) ;)

Reply:
acf68417